### PR TITLE
Add unified entity search endpoint

### LINKED ISSUE #60: Add unified entity search endpoint for @mentions and entity linking
### Problem

The application needs the ability to search for entities (users, shots, assets, versions, etc.) by name for two key use cases:

1. **@Mentions in notes**: While typing a note, users should be able to type `@` and search for people, shots, assets, and other entities to mention/link
2. **Notes properties panel**: Users need to search for and select entities to link notes to

The current `/find` endpoint requires:
- Separate API calls per entity type
- Manual construction of filter objects
- No result limiting for typeahead performance
- No project scoping by default

This currently blocks the notes properties and @ menitions features

### Proposed Solution

Add a new `POST /search` endpoint that provides a simplified, aggregated search interface optimized for autocomplete/typeahead use cases.

### API Design

**Endpoint:** `POST /search`

**Request Body:**

```json
{
  "query": "john",
  "entity_types": ["user", "shot", "asset", "version"],
  "project_id": 123,
  "limit": 10
}
```

| Field | Type | Required | Description |
|-------|------|----------|-------------|
| `query` | string | Yes | Text to search for (searches `name` field) |
| `entity_types` | string[] | Yes | Entity types to search: `user`, `shot`, `asset`, `version`, `task`, `playlist` |
| `project_id` | int | No | Scope results to a specific project (recommended for non-user entities) |
| `limit` | int | No | Max results per entity type (default: 10) |

**Response:**

```json
{
  "results": [
    {
      "type": "User",
      "id": 42,
      "name": "John Smith",
      "email": "john.smith@studio.com"
    },
    {
      "type": "Shot",
      "id": 101,
      "name": "shot_010_0020",
      "description": "Hero enters frame"
    },
    {
      "type": "Asset",
      "id": 55,
      "name": "johnny_rig",
      "description": "Main character rig"
    }
  ]
}
```

#### Backend

1. **New request/response models** in `backend/src/dna/models/requests.py`:
   - `SearchRequest` with query, entity_types, project_id, limit fields
   - `SearchResult` as a lightweight entity representation

2. **New endpoint** in `backend/src/main.py`:
   - `POST /search` handler that:
     - Validates entity types against allowed list
     - Calls `provider.find()` for each requested entity type in parallel
     - Applies `contains` filter on the name field
     - Applies project filter for non-user entities when `project_id` provided
     - Limits results per entity type
     - Aggregates and returns results

3. **Provider enhancement** (optional):
   - Add a `search()` method to `ProdtrackProviderBase` that handles multi-entity search in a single method
   - Could optimize by batching ShotGrid API calls where possible

#### Frontend (future issue)

- Add `searchEntities()` function to `@dna/core` API handler
- Create `useEntitySearch` hook for debounced search
- Build `EntitySearchDropdown` component for both @mentions and properties panel

### Acceptance Criteria

- [ ] `POST /search` endpoint accepts query, entity_types, project_id, and limit
- [ ] Returns results from all requested entity types in a single response
- [ ] Results are filtered by name containing the query string (case-insensitive)
- [ ] Non-user entities are scoped to project when project_id is provided
- [ ] Users are searchable without project scope (they're global)
- [ ] Results are limited to `limit` per entity type (default 10)
- [ ] Each result includes: type, id, name, and type-specific fields (email for users, description for shots/assets)
- [ ] Endpoint is documented in OpenAPI/Swagger
- [ ] Unit tests cover search across multiple entity types
- [ ] Test coverage remains at 90%+

### Example Usage

**Searching for @mentions while typing a note:**

```typescript
// User types "@joh"
const response = await api.search({
  query: "joh",
  entity_types: ["user", "shot", "asset"],
  project_id: currentProjectId,
  limit: 5
});
// Returns matching users, shots, and assets for dropdown
```

**Searching for entities to link in properties panel:**

```typescript
// User searches for a shot to link
const response = await api.search({
  query: "010",
  entity_types: ["shot"],
  project_id: currentProjectId,
  limit: 20
});
```

### Technical Notes

- The existing `find()` method in `ShotgridProvider` already supports the `contains` operator for text matching
- `FIELD_MAPPING` in `shotgrid.py` defines which fields are available per entity type
- The name field maps to different SG fields: `name` for users, `code` for shots/assets/versions

### Related

- Existing `/find` endpoint can remain for advanced/power-user queries
- Future: @mentions parsing and rendering in note content
- Future: Entity link display in notes properties panel

### PR DESCRIPTION
**Pull request by aabhinavvvvvvv**
# Add unified entity search endpoint Fixes #60 

## Summary

Implements a unified `POST /search` endpoint that enables typeahead searching across multiple entity types (user, shot, asset, version, task, playlist) in a single API call. This endpoint is required to unblock the @mentions feature (#61) and entity linking in the notes properties panel (#50).

### Models (`src/dna/models/requests.py`)
- Added `SearchRequest` model with `query`, `entity_types`, `project_id`, and `limit` fields
- Added `SearchResult` model for lightweight entity representation
- Includes type-specific fields: `email` for users, `description` for shots/assets/versions

### Provider Layer
- Added abstract `search()` method to `ProdtrackProviderBase` (`src/dna/prodtrack_providers/prodtrack_provider_base.py`)
- Implemented `search()` in `ShotgridProvider` (`src/dna/prodtrack_providers/shotgrid.py`)
  - Case-insensitive name filtering using "contains" operator
  - Optional project scoping for non-user entities
  - Configurable result limit per entity type (default: 10)
  - Returns lightweight dicts instead of full entity objects for performance

### API Endpoint (`src/main.py`)
- Added `POST /search` endpoint with request validation
- Validates entity types against `ENTITY_MODELS`
- Normalizes entity type casing (accepts "User", "USER", "user")
- Returns results in `{"results": [...]}` format
- Comprehensive error handling with proper HTTP status codes

### Tests (`tests/test_main.py`)
- Added `TestSearchEndpoint` class with 14 comprehensive test cases
- Tests cover success cases, validation, error handling, and edge cases
- 100% coverage for new `SearchRequest` and `SearchResult` models

## Test Coverage

All 14 new tests passing:

### CODE REVIEW

**Review by jspada200 — CHANGES_REQUESTED** [review]
When trying to search for the robot asset in SG I am getting a traceback. It might be the convert to DNA entity logic. If you are able to take a look as part of this PR, that would be great, otherwise I can take a look.

curl -X POST http://localhost:8000/search \
  -H "Content-Type: application/json" \
  -d '{
    "query": "rob", 
    "entity_types": ["shot", "asset", "version"]
  }'


For the Failed pipeline, just run `make format-python` and it should resolve the formatting issues.

**Comment by aabhinavvvvvvv** [comment]
> Another thing we may want to consider. At the moment we are using the `find` call that is abstracted. This queieries all the fields. Since we need something really snappy, we prob will want to only query the fields we need from within the search call (Type, ID, and code).

I've updated search() to bypass the abstracted find() entirely. Now calling self.sg.find() directly with only the minimal fields needed:

["id", "code"] for most entities
+ "email" for users
+ "description", "project" for shots/assets/versions

File: backend/src/dna/prodtrack_providers/shotgrid.py [hunk 1]
```diff
@@ -377,6 +381,107 @@ def find(self, entity_type: str, filters: list[dict[str, Any]]) -> list[EntityBa
...
+            sg_results = self.sg.find(
+                sg_entity_type,
+                filters=sg_filters,
+                fields=sg_fields,
+                limit=limit,
+            )
+
+            # Convert to lightweight search results directly from SG response
+            # Use DNA entity type instead of SG type
+            dna_type = entity_type.capitalize()
```
Comment (jspada200) [inline backend/src/dna/prodtrack_providers/shotgrid.py]: Should be a mapping, not just capitalize. For example, we will probably add a 'Location' entity in DNA. This will probably be somthing like `CustomEntity....` in SG.
